### PR TITLE
Ignore class properties in destructuring-assignment

### LINKED ISSUE #1875: destructuring-assignment in class properties
First I'd like to just say that I like the new `destructuring-assignment` rule because I think it enforces a lot more clarity and explicitness. However I'm having problems with it throwing errors inside of class properties. 

```javascript
class MyComponent extends React.Component {
  state = {
    value: this.props.defaultValue,
  };

  ...
}
```
IMO this should be an exception and be totally valid or at least have the option to exclude it just from this use-case. I know that I could move it inside a `constructor` and destructure it there but I'd rather not do that.

### PR DESCRIPTION
**Pull request by alexandernanberg**
Fixes #1875

Please let me know if you would like me to add more tests or if anything can be improved

### CODE REVIEW

**Comment by ljharb** [comment]
That's one of the points of the rule - if you need to rely on props values, it's supposed to force it to go in the constructor.

**Comment by alexandernanberg** [comment]
Okay I get it!

How would you like the option to be implemented? Not sure that I see a clear way to do it well without a breaking change. I guess we could add another enum option like `ignoreClassProperties` but I feel that it would maybe be a bit ambiguous?

**Comment by ljharb** [comment]
`allowInClassFields` seems reasonable to me, or `allow: ["class-fields"]` (where that's the only allowed value atm)

**Comment by alexandernanberg** [comment]
Okay so I've now added the option, I went for another name because I thought it was more logical than using the "allow" word, but let me know if you want me to change it. I also a bit unsure if I wrote the schema correctly

So you would do this;
`["error", "always", { "ignoreClassFields": true }]`

File: lib/rules/destructuring-assignment.js [hunk 1]
```diff
@@ -22,12 +22,20 @@ module.exports = {
...
+        }
+      },
+      additionalProperties: false
     }]
   },
 
   create: Components.detect((context, components, utils) => {
     const configuration = context.options[0] || DEFAULT_OPTION;
-
+    const ignoreClassFields = configuration === 'always' && context.options[1] && context.options[1].ignoreClassFields === true || false;
```
Comment (ljharb) [inline lib/rules/destructuring-assignment.js]: when set to “never”, this should still apply - you could maybe have `foo = function () { const { a } = this.props; }.call(this)`, or when do expressions land, you could have `foo = do { const { a } = this.props; a; }`

Comment (alexandernanberg) [reply lib/rules/destructuring-assignment.js]: Makes sense, I'll update my PR 👍

Comment (ljharb) [reply lib/rules/destructuring-assignment.js]: shouldn't https://github.com/yannickcr/eslint-plugin-react/pull/1909/files#diff-d73ce54079fb0715f60db27176a09fb6R135 also be updated?

**Comment by HenriBeck** [comment]
@ljharb should this be behind a flag because it doesn't have to do anything with destructuring assignments? I would agree that it would be a great rule to disallow props usage inside class properties but doing it here might be a little bit confusing for people.

File: lib/rules/destructuring-assignment.js (hunk 1 above)
Comment (alexandernanberg) [reply lib/rules/destructuring-assignment.js]: Yep it probably should. Sorry for missing that, I've never worked in AST or eslint-plugin projects before so it's all new to me. I've updated the PR again.

**Comment by ljharb** [comment]
@HenriBeck the rule is to force, or forbid, destructuring assignment regardless of location inside a class body. That class fields make destructuring awkward might be an inconvenience to some, hence the need for an option - but personally, I like the side effect that if you're trying to do anything complex in a class field, it forces you to properly put it in a constructor.

**Comment by HenriBeck** [comment]
@ljharb I agree with you that doing complicated things inside class properties isn't the best idea.
But forbidding it in this rule seems odd because class properties aren't about destructuring and should be able to disallow them whether or not you have this rule configured.

**Comment by ljharb** [comment]
Methods aren't about destructuring either - the rule deals with any references to state, props, or context, no matter where they appear.

**Comment by alexandernanberg** [comment]
Well shit, this does not work when assigning an object with properties containing `this.props` to a class field.

```js
class Foo extends React.Component {
  state = {
    // this will still throw an error
    bar: this.props.bar
  }
}
```

Did a quick try to fix this but didn't make any progress, can somebody either try to solve this or point me in the right direction? @ljharb

**Comment by EvHaus** [comment]
That code you posted isn't valid JS. You cannot use `this.props` in a class property definition. You'd have to rewrite this to:

```js
class Foo extends React.Component {
  constructor (props) {
    super(props);

    this.state = {
      // this will still throw an error
      bar: props.bar
    }
  }
}
```

**Comment by anna93** [comment]
Yup 7.11.1, the below still warns me about de-structuring props assignment:
 const sampleSchemaData = this.props.schemaData;

**Comment by ljharb** [comment]
@anna93 that’s a variable, not a class field.

**Comment by zslabs** [comment]
This still throws one for me using the same setup as @anna93:

```
class Input extends React.Component {
  id = `${this.props.name}-${generateUUID()}`;
 ...
```

**Comment by ljharb** [comment]
@zslabs if with the option enabled you still get a warning there, please file a new issue

**Pull request merged**